egs/wsj/s5/steps/nnet3/make_bottleneck_features.sh

```diff
@@ -0,0 +1,128 @@
+   echo "main options (for others, see top of script file)"
+   echo "  --config <config-file>                           # config containing options"
+   echo "  --nj <nj>                                        # number of parallel jobs"
+   echo "  --cmd (utils/run.pl|utils/queue.pl <queue opts>) # how to run jobs."
+   echo "  --ivector-dir                                    # directory for ivectors"
+   exit 1;
+fi
+
+data=$1
+bnf_data=$2
+nnetdir=$3
+archivedir=$4
+dir=$5
+
+# Assume that final.nnet is in nnetdir
+cmvn_opts=`cat $nnetdir/cmvn_opts`;
+bnf_nnet=$nnetdir/final.raw
+if [ ! -f $bnf_nnet ] ; then
+  echo "No such file $bnf_nnet";
+  exit 1;
+fi
+
+if $use_gpu; then
+  compute_queue_opt="--gpu 1"
+  compute_gpu_opt="--use-gpu=yes"
+  if ! cuda-compiled; then
+    echo "$0: WARNING: you are running with one thread but you have not compiled"
+    echo "   for CUDA.  You may be running a setup optimized for GPUs.  If you have"
+    echo "   GPUs and have nvcc installed, go to src/ and do ./configure; make"
+    exit 1
+  fi
+else
+  echo "$0: without using a GPU this will be very slow.  nnet3 does not yet support multiple threads."
+  compute_gpu_opt="--use-gpu=no"
+fi
+
+
+## Set up input features of nnet
+name=`basename $data`
+sdata=$data/split$nj
+
+mkdir -p $dir/log
+mkdir -p $bnf_data
+echo $nj > $nnetdir/num_jobs
+splice_opts=`cat $nnetdir/splice_opts 2>/dev/null`
+[[ -d $sdata && $data/feats.scp -ot $sdata ]] || split_data.sh $data $nj || exit 1;
+
+if [ "$ivector_dir" != "" ];then 
+  use_ivector=true
+fi
+
+feats="ark,s,cs:apply-cmvn $cmvn_opts --utt2spk=ark:$sdata/JOB/utt2spk scp:$sdata/JOB/cmvn.scp scp:$sdata/JOB/feats.scp ark:- |"
+ivec_feats="scp:utils/filter_scp.pl $sdata/JOB/utt2spk $ivector_dir/ivector_online.scp |"
+
+if [ ! -z "$transform_dir" ]; then
+  echo "Using transforms from $transform_dir"
+  [ ! -f $transform_dir/trans.1 ] && echo "No such file $transform_dir/trans.1" && exit 1;
+  transform_nj=`cat $transform_dir/num_jobs` || exit 1;
+  if [ "$nj" != "$transform_nj" ]; then
+    for n in $(seq $transform_nj); do cat $transform_dir/trans.$n; done >$dir/trans.ark
+    feats="$feats transform-feats --utt2spk=ark:$sdata/JOB/utt2spk ark,s,cs:$dir/trans.ark ark:- ark:- |"
+  else
+    feats="$feats transform-feats --utt2spk=ark:$sdata/JOB/utt2spk ark,s,cs:$transform_dir/trans.JOB ark:- ark:- |"
+  fi
+fi
+
+
+if [ $stage -le 1 ]; then
+  echo "$0: Generating bottle-neck features"
+  echo output-node name=output input=$bnf_name > output.config
+  modified_bnf_nnet="nnet3-copy --nnet-config=output.config $bnf_nnet - |"
+  ivector_opts=
+  if $use_ivector; then
+    ivec_period=`grep ivector-period $ivector_dir/conf/ivector_extractor.conf  | cut -d"=" -f2` 
+    ivector_opts="--online-ivector-period=$ivec_period --online-ivectors='$ivec_feats'"
+  fi
+  $cmd $compute_queue_opt JOB=1:$nj $dir/log/make_bnf_$name.JOB.log \
+    nnet3-compute $compute_gpu_opt $ivector_opts "$modified_bnf_nnet" "$feats" ark:- \| \
```

Comment:
I think you might want to make your bnf extractor architecture agnostic. If you are using an RNN you should support specification of extra-left-context, extra-right-context and chunk width.